fem/restriction.hpp

```diff
@@ -290,6 +298,16 @@ class ConformingFaceRestriction : public FaceRestriction
    void AddMultTranspose(const Vector &x, Vector &y,
                          const double a = 1.0) const override;
 
+   /** @brief Gather the degrees of freedom, i.e. goes from face E-Vector to
+       L-Vector @b not taking into account signs from DOF orientations.
+
+       @sa AddMultTranspose(). */
+   void AddMultTransposeUnsigned(const Vector &x, Vector &y,
+                                 const double a = 1.0) const override;
+
+   void AddMultTranspose(const Vector &x, Vector &y, bool use_signs,
```

Comment:
A `brief` comment would be useful, especially for this new `use_signs` showing at the interface level.

Reply:
This is what the free function would look like. What do you think?

```diff
diff --git a/fem/restriction.cpp b/fem/restriction.cpp
index 26d1ac7fd..ed203b8f8 100644
--- a/fem/restriction.cpp
+++ b/fem/restriction.cpp
@@ -675,36 +675,31 @@ void ConformingFaceRestriction::Mult(const Vector& x, Vector& y) const
    });
 }
 
-void ConformingFaceRestriction::AddMultTranspose(
-   const Vector& x, Vector& y, const double a) const
-{
-   AddMultTranspose(x, y, true, a);
-}
-
-void ConformingFaceRestriction::AddMultTransposeUnsigned(
-   const Vector& x, Vector& y, const double a) const
-{
-   AddMultTranspose(x, y, false, a);
-}
-
-void ConformingFaceRestriction::AddMultTranspose(
-   const Vector& x, Vector& y, bool use_signs, const double a) const
+static void ConformingFaceRestriction_AddMultTranspose(
+   const int ndofs,
+   const int face_dofs,
+   const int nf,
+   const int vdim,
+   const bool by_vdim,
+   const Array<int> &gather_offsets,
+   const Array<int> &gather_indices,
+   const Vector &x,
+   Vector &y,
+   bool use_signs,
+   const double a)
 {
    MFEM_VERIFY(a == 1.0, "General coefficient case is not yet supported!");
    if (nf==0) { return; }
    // Assumes all elements have the same number of dofs
-   const int nface_dofs = face_dofs;
-   const int vd = vdim;
-   const bool t = byvdim;
    auto d_offsets = gather_offsets.Read();
    auto d_indices = gather_indices.Read();
-   auto d_x = Reshape(x.Read(), nface_dofs, vd, nf);
-   auto d_y = Reshape(y.ReadWrite(), t?vd:ndofs, t?ndofs:vd);
+   auto d_x = Reshape(x.Read(), face_dofs, vdim, nf);
+   auto d_y = Reshape(y.ReadWrite(), by_vdim?vdim:ndofs, by_vdim?ndofs:vdim);
    MFEM_FORALL(i, ndofs,
    {
       const int offset = d_offsets[i];
       const int next_offset = d_offsets[i + 1];
-      for (int c = 0; c < vd; ++c)
+      for (int c = 0; c < vdim; ++c)
       {
          double dof_value = 0;
          for (int j = offset; j < next_offset; ++j)
@@ -712,13 +707,29 @@ void ConformingFaceRestriction::AddMultTranspose(
             const int s_idx_j = d_indices[j];
             const int sgn = (s_idx_j >= 0 || !use_signs) ? 1 : -1;
             const int idx_j = (s_idx_j >= 0) ? s_idx_j : -1 - s_idx_j;
-            dof_value += sgn*d_x(idx_j % nface_dofs, c, idx_j / nface_dofs);
+            dof_value += sgn*d_x(idx_j % face_dofs, c, idx_j / face_dofs);
          }
-         d_y(t?c:i,t?i:c) += dof_value;
+         d_y(by_vdim?c:i,by_vdim?i:c) += dof_value;
       }
    });
 }
 
+void ConformingFaceRestriction::AddMultTranspose(
+   const Vector& x, Vector& y, const double a) const
+{
+   ConformingFaceRestriction_AddMultTranspose(
+      ndofs, face_dofs, nf, vdim, byvdim, gather_offsets, gather_indices, x, y,
+      true, a);
+}
+
+void ConformingFaceRestriction::AddMultTransposeUnsigned(
+   const Vector& x, Vector& y, const double a) const
+{
+   ConformingFaceRestriction_AddMultTranspose(
+      ndofs, face_dofs, nf, vdim, byvdim, gather_offsets, gather_indices, x, y,
+      false, a);
+}
+
 void ConformingFaceRestriction::CheckFESpace(const ElementDofOrdering
                                              f_ordering)
 {
diff --git a/fem/restriction.hpp b/fem/restriction.hpp
index ce9e9ca3c..50c001615 100644
--- a/fem/restriction.hpp
+++ b/fem/restriction.hpp
@@ -305,9 +305,6 @@ public:
    void AddMultTransposeUnsigned(const Vector &x, Vector &y,
                                  const double a = 1.0) const override;
 
-   void AddMultTranspose(const Vector &x, Vector &y, bool use_signs,
-                         const double a) const;
-
 private:
    /** @brief Compute the scatter indices: L-vector to E-vector, and the offsets
        for the gathering: E-vector to L-vector.
```

Reply:
I see - I'd be in favor to hide this logic from the user and use as you suggest the free function, but that can be discussed.

Reply:
Fine by me. Done in 87c792899f48cb4fefae5de5788cc63d727996d1